### PR TITLE
[AIRFLOW-5269] Reuse session in Scheduler Job from health endpoint

### PR DESCRIPTION
**Pull request by kaxil**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5269

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
Reuse session in Scheduler Job from health endpoint

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Current tests cover it

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Review by milton0825 — APPROVED** [review]

**Review by Fokko — CHANGES_REQUESTED** [review]
I would remove the decorator from the whole `/health` endpoint:
```python
class Airflow(AirflowBaseView):
    @expose('/health')
    @provide_session
    def health(self, session=None):
        """
        An endpoint helping check the health status of the Airflow instance,
        including metadatabase and scheduler.
        """

        payload = {
            'metadatabase': {'status': 'unhealthy'}
        }

        latest_scheduler_heartbeat = None
        scheduler_status = 'unhealthy'
        payload['metadatabase'] = {'status': 'healthy'}
        try:
            scheduler_job = jobs.SchedulerJob.most_recent_job(session=session)

            if scheduler_job:
                latest_scheduler_heartbeat = scheduler_job.latest_heartbeat.isoformat()
                if scheduler_job.is_alive():
                    scheduler_status = 'healthy'
        except Exception:
            payload['metadatabase']['status'] = 'unhealthy'

        payload['scheduler'] = {'status': scheduler_status,
                                'latest_scheduler_heartbeat': latest_scheduler_heartbeat}

        return wwwutils.json_response(payload)
```
We don't use the session anywhere in the function. This will only keep the session open inside the scope of the `most_recent_job()` function. WDYT @kaxil ?

**Comment by kaxil** [comment]
@Fokko Updated

**Review by Fokko — APPROVED** [review]
LGTM! 👍

File: airflow/jobs/base_job.py [hunk 1]
```diff
@@ -93,7 +93,7 @@ def __init__(
 
     @classmethod
     @provide_session
-    def most_recent_job(cls, session: Session) -> Optional['BaseJob']:
+    def most_recent_job(cls, session=None) -> Optional['BaseJob']:
```
Comment (ashb) [inline airflow/jobs/base_job.py:96]: Did we need to change this to `None`? This is  quirk between type checkers and docs here. :/

Comment (kaxil) [reply airflow/jobs/base_job.py:96]: Unfortunately yes, the below line calls it without a session.

https://github.com/apache/airflow/blob/aa8a30099884efbf6fd5fb276f3dc6a310866aba/airflow/www/views.py#L184

Comment (ashb) [reply airflow/jobs/base_job.py:96]: I mean it looks like it does, but the decorator takes care of that.

So interaction between our provide_session and typecheckers then.

**Pull request merged**